diff --git a/irq-gic.c.org b/irq-gic.c
index 412196a..db50b4f 100644
--- a/drivers/irqchip/irq-gic.c
+++ b/drivers/irqchip/irq-gic.c
@@ -466,23 +466,39 @@ static void gic_cpu_if_up(struct gic_chip_data *gic)
 
 static void gic_dist_init(struct gic_chip_data *gic)
 {
-	unsigned int i;
-	u32 cpumask;
+	unsigned int i, j, k;
+	u32 cpumask0, cpumask, no_init_num, no_init_intno = 0;	
 	unsigned int gic_irqs = gic->gic_irqs;
 	void __iomem *base = gic_data_dist_base(gic);
 
-	writel_relaxed(GICD_DISABLE, base + GIC_DIST_CTRL);
+	/* writel_relaxed(GICD_DISABLE, base + GIC_DIST_CTRL); */
 
 	/*
 	 * Set all global interrupts to this CPU only.
 	 */
-	cpumask = gic_get_cpumask(gic);
-	cpumask |= cpumask << 8;
-	cpumask |= cpumask << 16;
-	for (i = 32; i < gic_irqs; i += 4)
-		writel_relaxed(cpumask, base + GIC_DIST_TARGET + i * 4 / 4);
+	if (of_property_read_u32(irq_domain_get_of_node(gic->domain), "#noinit-spi-interrupt-cells", &no_init_num))
+		no_init_num = 0;
+
+	j = 0; 
+	cpumask0 = gic_get_cpumask(gic);
+	for (i = 32; i < gic_irqs; i += 4) {
+		cpumask = readl(base + GIC_DIST_TARGET + i * 4 / 4);
+		for (k = 0; k < 4; k++) {
+			if (no_init_num > j) {
+				if (of_property_read_u32_index(irq_domain_get_of_node(gic->domain), "noinit-spi-interrupts", j, &no_init_intno))
+					return;
+			}
+			if ((i + k) == no_init_intno) {
+				j++;
+			} else {
+				cpumask &= ~(0x000000FF << (k*8));
+				cpumask |= (cpumask0   << (k*8));
+			}
+		}
+ 		writel_relaxed(cpumask, base + GIC_DIST_TARGET + i * 4 / 4);
+	}
 
-	gic_dist_config(base, gic_irqs, NULL);
+	gic_dist_config(gic->domain, base, gic_irqs, NULL);
 
 	writel_relaxed(GICD_ENABLE, base + GIC_DIST_CTRL);
 }
@@ -868,11 +884,11 @@ static __init void gic_smp_init(void)
 				  "irqchip/arm/gic:starting",
 				  gic_starting_cpu, NULL);
 
-	base_sgi = irq_domain_alloc_irqs(gic_data[0].domain, 8, NUMA_NO_NODE, &sgi_fwspec);
+	base_sgi = irq_domain_alloc_irqs(gic_data[0].domain, 11, NUMA_NO_NODE, &sgi_fwspec);
 	if (WARN_ON(base_sgi <= 0))
 		return;
 
-	set_smp_ipi_range(base_sgi, 8);
+	set_smp_ipi_range(base_sgi, 11);
 }
 #else
 #define gic_smp_init()		do { } while(0)
