diff --git a/irq-gic-v3.c.org b/irq-gic-v3.c
index a09883e..36eb0b6 100644
--- a/drivers/irqchip/irq-gic-v3.c
+++ b/drivers/irqchip/irq-gic-v3.c
@@ -923,7 +923,8 @@ static bool gic_has_group0(void)
 
 static void __init gic_dist_init(void)
 {
-	unsigned int i;
+	unsigned int i, j;
+	u32 no_init_num, no_init_intno = 0;
 #ifndef CONFIG_GIC_GENTLE_CONFIG
 	u64 affinity;
 #endif
@@ -944,8 +945,11 @@ static void __init gic_dist_init(void)
        }
 #endif
 
+	if (of_property_read_u32(irq_domain_get_of_node(gic_data.domain), "#noinit-spi-interrupt-cells", &no_init_num))
+		no_init_num = 0;
+
 	/* Disable the distributor */
-	writel_relaxed(0, base + GICD_CTLR);
+	/* writel_relaxed(0, base + GICD_CTLR); */
 	gic_dist_wait_for_rwp();
 
 	/*
@@ -973,7 +977,7 @@ static void __init gic_dist_init(void)
 		writel_relaxed(GICD_INT_DEF_PRI_X4, base + GICD_IPRIORITYRnE + i);
 
 	/* Now do the common stuff */
-	gic_dist_config(base, GIC_LINE_NR, NULL);
+	gic_dist_config(gic_data.domain, base, GIC_LINE_NR, NULL);
 
 	val = GICD_CTLR_ARE_NS | GICD_CTLR_ENABLE_G1A | GICD_CTLR_ENABLE_G1;
 	if (gic_data.rdists.gicd_typer2 & GICD_TYPER2_nASSGIcap) {
@@ -998,8 +1002,18 @@ static void __init gic_dist_init(void)
 	 * enabled.
 	 */
 	affinity = gic_cpu_to_affinity(smp_processor_id());
-	for (i = 32; i < GIC_LINE_NR; i++)
-		gic_write_irouter(affinity, base + GICD_IROUTER + i * 8);
+	j = 0;
+	for (i = 32; i < GIC_LINE_NR; i++) {
+		if (no_init_num > j) {
+			if (of_property_read_u32_index(irq_domain_get_of_node(gic_data.domain), "noinit-spi-interrupts", j, &no_init_intno))
+				return;
+		}
+		if (i == no_init_intno) {
+			j++;
+		} else {
+			gic_write_irouter(affinity, base + GICD_IROUTER + i * 8);
+		}
+	}
 
 	for (i = 0; i < GIC_ESPI_NR; i++)
 		gic_write_irouter(affinity, base + GICD_IROUTERnE + i * 8);
@@ -1431,11 +1445,11 @@ static void __init gic_smp_init(void)
 				  gic_starting_cpu, NULL);
 
 	/* Register all 8 non-secure SGIs */
-	base_sgi = irq_domain_alloc_irqs(gic_data.domain, 8, NUMA_NO_NODE, &sgi_fwspec);
+	base_sgi = irq_domain_alloc_irqs(gic_data.domain, 11, NUMA_NO_NODE, &sgi_fwspec);
 	if (WARN_ON(base_sgi <= 0))
 		return;
 
-	set_smp_ipi_range(base_sgi, 8);
+	set_smp_ipi_range(base_sgi, 11);
 }
 
 static int gic_set_affinity(struct irq_data *d, const struct cpumask *mask_val,
