diff --git a/irq-gic-common.c.org b/irq-gic-common.c
index 100bef1..48735a7 100644
--- a/drivers/irqchip/irq-gic-common.c
+++ b/drivers/irqchip/irq-gic-common.c
@@ -90,23 +90,57 @@ int gic_configure_irq(unsigned int irq, unsigned int type,
 	return ret;
 }
 
-void gic_dist_config(void __iomem *base, int gic_irqs,
+void gic_dist_config(struct irq_domain	*domain, void __iomem *base, int gic_irqs,
 		     void (*sync_access)(void))
 {
-	unsigned int i;
+	unsigned int i, j, k;
+	u32 no_init_num, no_init_intno = 0;
+	u32 regvalue;
 
+	if (of_property_read_u32(irq_domain_get_of_node(domain), "#noinit-spi-interrupt-cells", &no_init_num))
+		no_init_num = 0;
 	/*
 	 * Set all global interrupts to be level triggered, active low.
 	 */
-	for (i = 32; i < gic_irqs; i += 16)
-		writel_relaxed(GICD_INT_ACTLOW_LVLTRIG,
-					base + GIC_DIST_CONFIG + i / 4);
+	j = 0;
+	for (i = 32; i < gic_irqs; i += 16) {
+		regvalue = readl(base + GIC_DIST_CONFIG + i / 4);
+		for (k = 0; k < 16; k++) {
+			if (no_init_num > j) {
+				if (of_property_read_u32_index(irq_domain_get_of_node(domain), "noinit-spi-interrupts", j, &no_init_intno)) 
+					return;
+			}
+			if ((i + k) == no_init_intno) {
+				j++;
+			} else {
+				regvalue &= ~(0x00000003 << (k*2));
+				regvalue |= ((GICD_INT_ACTLOW_LVLTRIG & 0x00000003) << (k*2));
+			}
+		}
+		writel_relaxed(regvalue,
+ 					base + GIC_DIST_CONFIG + i / 4);
+	}
 
 	/*
 	 * Set priority on all global interrupts.
 	 */
-	for (i = 32; i < gic_irqs; i += 4)
-		writel_relaxed(GICD_INT_DEF_PRI_X4, base + GIC_DIST_PRI + i);
+	j = 0;
+	for (i = 32; i < gic_irqs; i += 4) {
+		regvalue = readl(base + GIC_DIST_PRI + i);
+		for (k = 0; k < 4; k++) {
+			if (no_init_num > j) {
+				if (of_property_read_u32_index(irq_domain_get_of_node(domain), "noinit-spi-interrupts", j, &no_init_intno)) 
+					return;
+			}
+			if ((i + k) == no_init_intno) {
+				j++;
+			} else {
+				regvalue &= ~(0x000000FF << (k*8));
+				regvalue |= (GICD_INT_DEF_PRI << (k*8));
+			}
+		}
+		writel_relaxed(regvalue, base + GIC_DIST_PRI + i);
+	}
 
 #ifndef CONFIG_GIC_GENTLE_CONFIG
 	/*
@@ -120,11 +154,24 @@ void gic_dist_config(void __iomem *base, int gic_irqs,
 	 * Deactivate and disable all SPIs. Leave the PPI and SGIs
 	 * alone as they are in the redistributor registers on GICv3.
 	 */
-	for (i = 32; i < gic_irqs; i += 32) {
-		writel_relaxed(GICD_INT_EN_CLR_X32,
-			       base + GIC_DIST_ACTIVE_CLEAR + i / 8);
-		writel_relaxed(GICD_INT_EN_CLR_X32,
-			       base + GIC_DIST_ENABLE_CLEAR + i / 8);
+	j = 0;
+ 	for (i = 32; i < gic_irqs; i += 32) {
+		regvalue = 0;
+		for (k = 0; k < 32; k++) {
+			if (no_init_num > j) {
+				if (of_property_read_u32_index(irq_domain_get_of_node(domain), "noinit-spi-interrupts", j, &no_init_intno)) 
+					return;
+			}
+			if ((i + k) == no_init_intno) {
+				j++;
+			} else {
+				regvalue |= (1 << k);
+			}
+		}
+		writel_relaxed(regvalue,
+					base + GIC_DIST_ACTIVE_CLEAR + i / 8);
+		writel_relaxed(regvalue,
+					base + GIC_DIST_ENABLE_CLEAR + i / 8);
 	}
 #endif
 
