diff --git a/uio.c.org b/uio.c
index 6ca4572..275319b 100644
--- a/drivers/uio/uio.c
+++ b/drivers/uio/uio.c
@@ -740,8 +740,14 @@ static int uio_mmap_physical(struct vm_area_struct *vma)
 		return -EINVAL;
 
 	vma->vm_ops = &uio_physical_vm_ops;
-	if (idev->info->mem[mi].memtype == UIO_MEM_PHYS)
-		vma->vm_page_prot = pgprot_noncached(vma->vm_page_prot);
+	if (idev->info->mem[mi].memtype == UIO_MEM_PHYS) {
+		if (mem->cache_attr == 1) {
+			vma->vm_page_prot = pgprot_writecombine(vma->vm_page_prot);
+		} else if (mem->cache_attr == 2) {
+			vma->vm_page_prot = pgprot_dmacoherent(vma->vm_page_prot);
+		} else if (mem->cache_attr == 0)
+			vma->vm_page_prot = pgprot_noncached(vma->vm_page_prot);
+	}
 
 	/*
 	 * We cannot use the vm_iomap_memory() helper here,
