--- ./drivers/uio/uio_pdrv_genirq.c.org	2021-10-18 17:46:00.074073499 +0900
+++ ./drivers/uio/uio_pdrv_genirq.c	2021-10-19 15:54:29.000000000 +0900
@@ -33,6 +33,7 @@
 	spinlock_t lock;
 	unsigned long flags;
 	struct platform_device *pdev;
+	unsigned int ipino;	
 };
 
 /* Bits in uio_pdrv_genirq_platdata.flags */
@@ -40,6 +41,15 @@
 	UIO_IRQ_DISABLED = 0,
 };
 
+static struct uio_info* ipi_uio_info[16];
+int set_ipi_handler(int ipinr, void *handler, char *desc);
+void clear_ipi_handler(int ipinr);
+
+static void ipi_handler(int ipinr)
+{
+	uio_event_notify(ipi_uio_info[ipinr]);
+}
+
 static int uio_pdrv_genirq_open(struct uio_info *info, struct inode *inode)
 {
 	struct uio_pdrv_genirq_platdata *priv = info->priv;
@@ -103,8 +113,12 @@
 static void uio_pdrv_genirq_cleanup(void *data)
 {
 	struct device *dev = data;
-
-	pm_runtime_disable(dev);
+	struct uio_info *uioinfo = dev_get_platdata(dev);
+ 
+ 	pm_runtime_disable(dev);
+	if (((struct uio_pdrv_genirq_platdata *)uioinfo->priv)->ipino) {
+		clear_ipi_handler(((struct uio_pdrv_genirq_platdata *)uioinfo->priv)->ipino);
+	}
 }
 
 static int uio_pdrv_genirq_probe(struct platform_device *pdev)
@@ -114,6 +128,7 @@
 	struct uio_pdrv_genirq_platdata *priv;
 	struct uio_mem *uiomem;
 	int ret = -EINVAL;
+	const void *of_prop;
 	int i;
 
 	if (node) {
@@ -170,6 +185,14 @@
 			dev_err(&pdev->dev, "failed to get IRQ\n");
 			return ret;
 		}
+
+		of_prop = of_get_property(pdev->dev.of_node, "ipi", NULL);
+		if (of_prop) {
+			uioinfo->irq = UIO_IRQ_CUSTOM;
+			priv->ipino = be32_to_cpup(of_prop);
+			ipi_uio_info[priv->ipino] = priv->uioinfo;
+			set_ipi_handler(priv->ipino, ipi_handler, "Kick From uC3");
+		}
 	}
 
 	if (uioinfo->irq) {
@@ -204,6 +227,13 @@
 			break;
 		}
 
+		of_prop = of_get_property(pdev->dev.of_node, "cache_attr", NULL);
+		if (of_prop) {
+			uiomem->cache_attr = be32_to_cpup(of_prop);
+		} else {
+			uiomem->cache_attr = 0;
+		}
+
 		uiomem->memtype = UIO_MEM_PHYS;
 		uiomem->addr = r->start & PAGE_MASK;
 		uiomem->offs = r->start & ~PAGE_MASK;
